cpp/src/arrow/python/python_to_arrow.h

```diff
@@ -18,18 +18,23 @@
 #ifndef ARROW_PYTHON_PYTHON_TO_ARROW_H
 #define ARROW_PYTHON_PYTHON_TO_ARROW_H
 
-#include "arrow/python/platform.h"
-
 #include <memory>
 #include <vector>
 
-#include "arrow/python/common.h"
-#include "arrow/python/pyarrow.h"
 #include "arrow/status.h"
 #include "arrow/util/visibility.h"
 
+// Forward declaring PyObject, see
+// https://mail.python.org/pipermail/python-dev/2003-August/037601.html
```

Comment:
@pitrou could you have a look at this?

Reply:
This header is included in the compilation of the TensorFlow Op now, which is done after the wheels have been created, on the user's computer (we decided about this strategy in https://github.com/apache/arrow/pull/2046, see the comments from @ppwwyyxx and @xhochy). On the user's computer, we don't have a fancy build environment (no CMake, no Python paths, etc.). This change makes it possible to compile the TensorFlow Op without a Python environment.

Reply:
Hmm, so my question is: why does the "TensorFlow Op" (I'm not sure what that is, sorry: is it some kind of computation kernel?) need to include the Arrow Python files?

Reply:
Yeah, it's a TensorFlow extension (a shared library) that can be loaded into TensorFlow and provides custom operators, like this one which reads/writes neural network weights from and to plasma efficiently.

It needs to include the Arrow Python files, because they define a serialization format for plasma that makes it possible to read/write objects from python. We use this in Ray to communicate TensorFlow weights between python and the TensorFlow runtime in an efficient way.

Making it so python_to_arrow.h and arrow_to_python.h don't depend on Python headers might be valuable in other contexts too.

Reply:
Unless the serialization format is Python-specific (is it, and if so, why?), perhaps it can be defined outside of `arrow/python`.